File: backend/routes/newsletter.js
/**
 * routes/newsletter.js — Subscribe / unsubscribe (MongoDB + JSON fallback)
 */
const express = require('express');
const { isConnected } = require('../db/mongoose');

const router = express.Router();
function getSub() { return require('../db/models').Subscriber; }
const db = () => require('../db/db');

// ── POST /api/newsletter/subscribe ────────────────────────────────────────────
router.post('/subscribe', async (req, res) => {
  try {
    const { email } = req.body;
    if (!email?.includes('@')) return res.status(400).json({ error: 'Valid email is required' });

    if (isConnected()) {
      await getSub().findOneAndUpdate(
        { email: email.toLowerCase().trim() },
        { email: email.toLowerCase().trim() },
        { upsert: true, new: true }
      );
    } else {
      const store = db();
      const exists = store.findOne('subscribers', s => s.email === email.toLowerCase().trim());
      if (!exists) store.push('subscribers', { email: email.toLowerCase().trim(), createdAt: new Date().toISOString() });
    }

    res.json({ message: '🌱 Subscribed! Welcome to the EcoPlant family. Expect weekly plant love in your inbox.' });
  } catch (err) {
    if (err.code === 11000) return res.json({ message: '🌱 Already subscribed! Thank you for your enthusiasm.' });
    console.error('[newsletter/subscribe]', err);
    res.status(500).json({ error: err.message });
  }
});

// ── POST /api/newsletter/unsubscribe ──────────────────────────────────────────
router.post('/unsubscribe', async (req, res) => {
  try {
    const { email } = req.body;
    if (!email) return res.status(400).json({ error: 'Email is required' });

    if (isConnected()) {
      await getSub().deleteOne({ email: email.toLowerCase().trim() });
    } else {
      db().remove('subscribers', s => s.email === email.toLowerCase().trim());
    }
    res.json({ message: 'Unsubscribed successfully.' });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// ── GET /api/newsletter/subscribers — Admin only ──────────────────────────────
router.get('/subscribers', async (req, res) => {
  const adminSecret = req.headers['x-admin-secret'];
  if (!adminSecret || adminSecret !== process.env.ADMIN_SECRET)
    return res.status(403).json({ error: 'Forbidden' });

  try {
    let subscribers;
    if (isConnected()) {
      subscribers = await getSub().find().sort({ createdAt: -1 }).lean();
    } else {
      subscribers = db().get('subscribers');
    }
    res.json({ subscribers, total: subscribers.length });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: python-backend/routers/newsletter.py
"""
routers/newsletter.py — Subscribe / unsubscribe
"""

import os
from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from db.database import is_connected, get_db, json_db

router = APIRouter(prefix="/api/newsletter", tags=["newsletter"])


class EmailBody(BaseModel):
    email: str


@router.post("/subscribe")
async def subscribe(body: EmailBody):
    if "@" not in body.email:
        raise HTTPException(400, "Valid email is required")
    email = body.email.lower().strip()

    if is_connected():
        await get_db().subscribers.update_one(
            {"email": email}, {"$set": {"email": email, "createdAt": datetime.now(timezone.utc)}},
            upsert=True,
        )
    else:
        existing = json_db.find_one("subscribers", lambda s: s["email"] == email)
        if not existing:
            json_db.push("subscribers", {"email": email, "createdAt": datetime.now(timezone.utc).isoformat()})

    return {"message": "🌱 Subscribed! Welcome to the EcoPlant family. Expect weekly plant love in your inbox."}


@router.post("/unsubscribe")
async def unsubscribe(body: EmailBody):
    if not body.email:
        raise HTTPException(400, "Email is required")
    email = body.email.lower().strip()

    if is_connected():
        await get_db().subscribers.delete_one({"email": email})
    else:
        json_db.remove("subscribers", lambda s: s["email"] == email)

    return {"message": "Unsubscribed successfully."}


@router.get("/subscribers")
async def get_subscribers(request: Request):
    admin_secret = request.headers.get("x-admin-secret", "")
    if admin_secret != os.getenv("ADMIN_SECRET", ""):
        raise HTTPException(403, "Forbidden")

    if is_connected():
        cursor = get_db().subscribers.find().sort("createdAt", -1)
        subs = []
        async for doc in cursor:
            doc["id"] = str(doc.pop("_id"))
            subs.append(doc)
    else:
        subs = json_db.get("subscribers")

    return {"subscribers": subs, "total": len(subs)}
